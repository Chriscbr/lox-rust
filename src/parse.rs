use crate::{
    ast::{Binary, BinaryOp, Expr, Grouping, Literal, Stmt, Unary, UnaryOp, VarDecl, Variable},
    error::Error,
    token::{Token, TokenType},
};

pub struct Parser {
    tokens: Vec<Token>,
    current: usize,
    errors: Vec<Error>,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Self {
            tokens,
            current: 0,
            errors: vec![],
        }
    }

    pub fn parse(&mut self) -> Result<Vec<Stmt>, Error> {
        let mut stmts = vec![];
        while !self.is_at_end() {
            let stmt = self.declaration();
            match stmt {
                Ok(stmt) => stmts.push(stmt),
                Err(err) => {
                    self.errors.push(err);
                    self.synchronize();
                }
            }
        }

        Ok(stmts)
    }

    fn expr(&mut self) -> Result<Expr, Error> {
        self.equality()
    }

    fn declaration(&mut self) -> Result<Stmt, Error> {
        if self.matches(&[TokenType::Var]) {
            return self.var_declaration();
        }

        self.stmt()
    }

    fn stmt(&mut self) -> Result<Stmt, Error> {
        if self.matches(&[TokenType::Print]) {
            return self.print_statement();
        }

        self.expr_stmt()
    }

    fn print_statement(&mut self) -> Result<Stmt, Error> {
        let expr = self.expr()?;
        if !self.check(TokenType::Semicolon) {
            return Err(Error::ExpectSemicolonAfterExpression {
                token: self.peek().clone(),
            });
        }
        self.advance();
        Ok(Stmt::Print(expr))
    }

    fn var_declaration(&mut self) -> Result<Stmt, Error> {
        if !self.check(TokenType::Identifier) {
            return Err(Error::ExpectedVariableName {
                token: self.peek().clone(),
            });
        }
        let name = self.advance().lexeme.clone();
        let initializer = if self.matches(&[TokenType::Equal]) {
            Some(self.expr()?)
        } else {
            None
        };

        if !self.check(TokenType::Semicolon) {
            return Err(Error::ExpectSemicolonAfterExpression {
                token: self.peek().clone(),
            });
        }
        self.advance();
        Ok(Stmt::VarDecl(VarDecl { name, initializer }))
    }

    fn expr_stmt(&mut self) -> Result<Stmt, Error> {
        let expr = self.expr()?;
        if !self.check(TokenType::Semicolon) {
            return Err(Error::ExpectSemicolonAfterExpression {
                token: self.peek().clone(),
            });
        }
        self.advance();
        Ok(Stmt::Expr(expr))
    }

    fn equality(&mut self) -> Result<Expr, Error> {
        let mut expr = self.comparison()?;

        while self.matches(&[TokenType::BangEqual, TokenType::EqualEqual]) {
            let op = self.previous().ty;
            let right = self.comparison()?;
            expr = Expr::Binary(Binary {
                left: Box::new(expr),
                op: BinaryOp::from(op),
                right: Box::new(right),
            });
        }

        Ok(expr)
    }

    fn comparison(&mut self) -> Result<Expr, Error> {
        let mut expr = self.term()?;

        while self.matches(&[
            TokenType::Greater,
            TokenType::GreaterEqual,
            TokenType::Less,
            TokenType::LessEqual,
        ]) {
            let op = self.previous().ty;
            let right = self.term()?;
            expr = Expr::Binary(Binary {
                left: Box::new(expr),
                op: BinaryOp::from(op),
                right: Box::new(right),
            });
        }

        Ok(expr)
    }

    fn term(&mut self) -> Result<Expr, Error> {
        let mut expr = self.factor()?;

        while self.matches(&[TokenType::Minus, TokenType::Plus]) {
            let op = self.previous().ty;
            let right = self.factor()?;
            expr = Expr::Binary(Binary {
                left: Box::new(expr),
                op: BinaryOp::from(op),
                right: Box::new(right),
            });
        }

        Ok(expr)
    }

    fn factor(&mut self) -> Result<Expr, Error> {
        let mut expr = self.unary()?;

        while self.matches(&[TokenType::Slash, TokenType::Star]) {
            let op = self.previous().ty;
            let right = self.unary()?;
            expr = Expr::Binary(Binary {
                left: Box::new(expr),
                op: BinaryOp::from(op),
                right: Box::new(right),
            });
        }

        Ok(expr)
    }

    fn unary(&mut self) -> Result<Expr, Error> {
        if self.matches(&[TokenType::Bang, TokenType::Minus]) {
            let op = self.previous().ty;
            let right = self.unary()?;
            return Ok(Expr::Unary(Unary {
                op: UnaryOp::from(op),
                right: Box::new(right),
            }));
        }

        return self.primary();
    }

    fn primary(&mut self) -> Result<Expr, Error> {
        if self.matches(&[TokenType::False]) {
            return Ok(Expr::Literal(Literal::Bool(false)));
        }

        if self.matches(&[TokenType::True]) {
            return Ok(Expr::Literal(Literal::Bool(true)));
        }

        if self.matches(&[TokenType::Nil]) {
            return Ok(Expr::Literal(Literal::Nil));
        }

        if self.matches(&[TokenType::Number]) {
            let value = str::parse::<f64>(&self.previous().lexeme).expect("invalid float");
            return Ok(Expr::Literal(Literal::Number(value)));
        }

        if self.matches(&[TokenType::String]) {
            let s = &self.previous().lexeme;
            // Trim the surrounding quotes.
            let trimmed = &s[1..s.len() - 1];
            return Ok(Expr::Literal(Literal::String(trimmed.to_string())));
        }

        if self.matches(&[TokenType::Identifier]) {
            return Ok(Expr::Variable(Variable {
                name: self.previous().lexeme.clone(),
            }));
        }

        if self.matches(&[TokenType::LeftParen]) {
            let expr = self.expr()?;
            if !self.check(TokenType::RightParen) {
                return Err(Error::ExpectRightParenAfterExpr {
                    token: self.peek().clone(),
                });
            }
            self.advance();
            return Ok(Expr::Grouping(Grouping {
                expr: Box::new(expr),
            }));
        }

        todo!("expected expression");
    }

    fn matches(&mut self, tys: &[TokenType]) -> bool {
        for ty in tys {
            if self.check(*ty) {
                self.advance();
                return true;
            }
        }

        false
    }

    fn check(&self, ty: TokenType) -> bool {
        if self.is_at_end() {
            return false;
        }

        return self.peek().ty == ty;
    }

    fn advance(&mut self) -> &Token {
        if !self.is_at_end() {
            self.current += 1;
        }

        return self.previous();
    }

    fn is_at_end(&self) -> bool {
        matches!(self.peek().ty, TokenType::EOF)
    }

    fn peek(&self) -> &Token {
        &self.tokens[self.current]
    }

    fn previous(&self) -> &Token {
        &self.tokens[self.current - 1]
    }

    fn synchronize(&mut self) {
        while !self.is_at_end() {
            if self.previous().ty == TokenType::Semicolon {
                return;
            }

            match self.peek().ty {
                TokenType::Class
                | TokenType::Fun
                | TokenType::Var
                | TokenType::For
                | TokenType::If
                | TokenType::While
                | TokenType::Print
                | TokenType::Return => return,
                _ => {}
            }

            self.advance();
        }
    }
}
