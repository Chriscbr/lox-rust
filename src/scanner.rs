use crate::{
    error::Error,
    token::{Token, TokenType, KEYWORDS},
};

#[derive(Debug)]
pub struct Scanner {
    source: String, // chars: Vec<char> ?
    tokens: Vec<Token>,
    start: usize,
    current: usize,
    line: usize,
    errors: Vec<Error>,
}

impl Scanner {
    pub fn new(source: String) -> Self {
        Self {
            source,
            tokens: vec![],
            start: 0,
            current: 0,
            line: 1,
            errors: vec![],
        }
    }

    pub fn scan_tokens(mut self) -> (Vec<Token>, Vec<Error>) {
        while !self.is_at_end() {
            // We are at the beginning of the next lexeme.
            self.start = self.current;
            self.scan_token();
        }

        self.tokens
            .push(Token::new(TokenType::EOF, "".to_string(), self.line));
        (self.tokens, self.errors)
    }

    fn is_at_end(&self) -> bool {
        return self.current >= self.source.len();
    }

    fn scan_token(&mut self) {
        let c = self.advance();
        match c {
            '(' => self.add_token(TokenType::LeftParen),
            ')' => self.add_token(TokenType::RightParen),
            '{' => self.add_token(TokenType::LeftBrace),
            '}' => self.add_token(TokenType::RightBrace),
            ',' => self.add_token(TokenType::Comma),
            '.' => self.add_token(TokenType::Dot),
            '-' => self.add_token(TokenType::Minus),
            '+' => self.add_token(TokenType::Plus),
            ';' => self.add_token(TokenType::Semicolon),
            '*' => self.add_token(TokenType::Star),
            '!' => {
                let ty = if self.matches('=') {
                    TokenType::BangEqual
                } else {
                    TokenType::Bang
                };
                self.add_token(ty);
            }
            '=' => {
                let ty = if self.matches('=') {
                    TokenType::EqualEqual
                } else {
                    TokenType::Equal
                };
                self.add_token(ty);
            }
            '<' => {
                let ty = if self.matches('=') {
                    TokenType::LessEqual
                } else {
                    TokenType::Less
                };
                self.add_token(ty);
            }
            '>' => {
                let ty = if self.matches('=') {
                    TokenType::GreaterEqual
                } else {
                    TokenType::Greater
                };
                self.add_token(ty);
            }
            '/' => {
                if self.matches('/') {
                    // A comment goes until the end of the line.
                    while self.peek() != '\n' && !self.is_at_end() {
                        self.advance();
                    }
                } else {
                    self.add_token(TokenType::Slash);
                }
            }
            ' ' | '\r' | '\t' => {} // Ignore whitespace.
            '\n' => {
                self.line += 1;
            }
            '"' => {
                self.string();
            }
            _ => {
                if self.is_digit(c) {
                    self.number();
                } else if self.is_alpha(c) {
                    self.identifier();
                } else {
                    self.errors
                        .push(Error::UnexpectedCharacter { line: self.line });
                }
            }
        };
    }

    fn identifier(&mut self) {
        while self.is_alpha_numeric(self.peek()) {
            self.advance();
        }

        let text = &self.source[self.start..self.current];
        let ty = KEYWORDS.get(text);
        match ty {
            Some(ty) => self.add_token(*ty),
            None => self.add_token(TokenType::Identifier),
        }

        self.add_token(TokenType::Identifier)
    }

    fn number(&mut self) {
        while self.is_digit(self.peek()) {
            self.advance();
        }

        // Look for a fractional part.
        if self.peek() == '.' && self.is_digit(self.peek_next()) {
            // Consume the "."
            self.advance();

            while self.is_digit(self.peek()) {
                self.advance();
            }
        }

        self.add_token(TokenType::Number);
    }

    fn string(&mut self) {
        while self.peek() != '"' && !self.is_at_end() {
            if self.peek() == '\n' {
                self.line += 1;
            }
            self.advance();
        }

        if self.is_at_end() {
            self.errors
                .push(Error::UnterminatedString { line: self.line });
        }

        self.advance(); // The closing ".

        self.add_token(TokenType::String);
    }

    fn matches(&mut self, expected: char) -> bool {
        if self.is_at_end() {
            return false;
        }

        if self.char_at(self.current) != expected {
            return false;
        }

        self.current += 1;
        return true;
    }

    fn peek(&self) -> char {
        if self.is_at_end() {
            return '\0';
        }
        return self.char_at(self.current);
    }

    fn peek_next(&self) -> char {
        if self.current + 1 >= self.source.len() {
            return '\0';
        }
        return self.char_at(self.current + 1);
    }

    fn is_alpha(&self, c: char) -> bool {
        c.is_ascii_alphabetic() || c == '_'
    }

    fn is_alpha_numeric(&self, c: char) -> bool {
        self.is_alpha(c) || self.is_digit(c)
    }

    fn is_digit(&self, c: char) -> bool {
        c.is_ascii_digit()
    }

    fn advance(&mut self) -> char {
        let c = self.char_at(self.current);
        self.current += 1;
        c
    }

    fn add_token(&mut self, ty: TokenType) {
        let text = &self.source[self.start..self.current];
        self.tokens
            .push(Token::new(ty, text.to_string(), self.line)); // str -> String
    }

    fn char_at(&self, idx: usize) -> char {
        return char::from(self.source.as_bytes()[idx]); // unsafe conversion
    }
}
