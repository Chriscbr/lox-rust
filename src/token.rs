use phf::phf_map;

pub static KEYWORDS: phf::Map<&'static str, TokenType> = phf_map! {
    "and" => TokenType::And,
    "class" => TokenType::Class,
    "else" => TokenType::Else,
    "false" => TokenType::False,
    "for" => TokenType::For,
    "fun" => TokenType::Fun,
    "if" => TokenType::If,
    "nil" => TokenType::Nil,
    "or" => TokenType::Or,
    "print" => TokenType::Print,
    "return" => TokenType::Return,
    "super" => TokenType::Super,
    "this" => TokenType::This,
    "true" => TokenType::True,
    "var" => TokenType::Var,
    "while" => TokenType::While,
};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenType {
    // Single-character tokens.
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,

    // One or two character tokens.
    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,

    // Literals.
    Identifier,
    String,
    Number,

    // Keywords
    And,
    Class,
    Else,
    False,
    Fun,
    For,
    If,
    Nil,
    Or,
    Print,
    Return,
    Super,
    This,
    True,
    Var,
    While,

    EOF,
}

#[derive(Debug, Clone)]
pub struct Token {
    pub ty: TokenType,
    pub lexeme: String,
    pub line: usize,
}

impl Token {
    pub fn new(ty: TokenType, lexeme: String, line: usize) -> Self {
        Self { ty, lexeme, line }
    }
}
