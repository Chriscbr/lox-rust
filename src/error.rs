// TODO: separate lex errors, parse errors, etc.

use crate::token::{Token, TokenType};

#[derive(Debug)]
pub enum Error {
    UnexpectedCharacter { line: usize },
    UnterminatedString { line: usize },
    ExpectRightParenAfterExpr { token: Token },
    ExpectSemicolonAfterExpression { token: Token },
    ExpectedVariableName { token: Token },
}

pub fn report_errors(errors: &[Error]) {
    for error in errors {
        match error {
            Error::UnexpectedCharacter { line } => {
                println!("[line {}] Error: unexpected character.", line)
            }
            Error::UnterminatedString { line } => {
                println!("[line {}] Error: unterminated string.", line)
            }
            Error::ExpectRightParenAfterExpr { token } => match token.ty {
                TokenType::EOF => println!(
                    "[line {}] Error at end: Expected ')' after expression.",
                    token.line
                ),
                _ => println!(
                    "[line {}] Error at '{}': Expected ')' after expression.",
                    token.line, token.lexeme,
                ),
            },
            Error::ExpectSemicolonAfterExpression { token } => match token.ty {
                TokenType::EOF => println!(
                    "[line {}] Error at end: Expected ';' after expression.",
                    token.line
                ),
                _ => println!(
                    "[line {}] Error at '{}': Expected ';' after expression.",
                    token.line, token.lexeme,
                ),
            },
            Error::ExpectedVariableName { token } => match token.ty {
                TokenType::EOF => println!(
                    "[line {}] Error at end: Expected variable name.",
                    token.line
                ),
                _ => println!(
                    "[line {}] Error at '{}': Expected variable name.",
                    token.line, token.lexeme
                ),
            },
        }
    }
}
