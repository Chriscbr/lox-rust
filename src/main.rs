mod ast;
mod error;
mod parse;
mod scanner;
mod token;

use std::fs::read_to_string;

use clap::Parser as ClapParser;
use error::Error;
use parse::Parser;
use scanner::Scanner;
use token::TokenType;

/// Run a Lox program.
#[derive(ClapParser, Debug)]
#[command(version, about)]
struct Args {
    /// File path
    file: String,
}

fn run(source: String) {
    let scanner = Scanner::new(source);
    let (tokens, errors) = scanner.scan_tokens();

    if errors.len() > 0 {
        report_errors(&errors);
        return;
    }

    let mut parser = Parser::new(tokens);
    let parsed = parser.parse();

    match parsed {
        Ok(expr) => println!("{}", expr),
        Err(err) => report_errors(&[err]),
    };
}

fn report_errors(errors: &[Error]) {
    for error in errors {
        match error {
            Error::UnexpectedCharacter { line } => {
                println!("[line {}] Error: unexpected character.", line)
            }
            Error::UnterminatedString { line } => {
                println!("[line {}] Error: unterminated string.", line)
            }
            Error::ExpectRightParenAfterExpression { token } => match token.ty {
                TokenType::EOF => println!(
                    "[line {}] Error at end: Expected ')' after expression.",
                    token.line
                ),
                _ => println!(
                    "[line {}] Error at '{}': Expected ')' after expression.",
                    token.line, token.lexeme,
                ),
            },
        }
    }
}

fn main() {
    let args = Args::parse();
    let source_text = read_to_string(args.file).expect("unable to read file");
    run(source_text);
}
