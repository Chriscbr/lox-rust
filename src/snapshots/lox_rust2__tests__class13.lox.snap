---
source: src/main.rs
expression: contents
---
stdout:
Yum!
Slurp!

stderr:
